packages/schedule/src/Schedule.js

```diff
@@ -3,174 +3,327 @@
-// scheduling a postMessage which gets scheduled after paint. Within the
-// postMessage handler do as much work as possible until time + frame rate.
-// By separating the idle call into a separate event tick we ensure that
-// layout, paint and other browser work is counted against the available time.
-// The frame rate is dynamically adjusted.
+let priorityContext = Deferred;
+let isPerformingWork = false;
+
+let isHostCallbackScheduled = false;
+
+const hasNativePerformanceNow =
+  typeof performance === 'object' && typeof performance.now === 'function';
+
+let timeRemaining;
+if (hasNativePerformanceNow) {
+  timeRemaining = () => {
+    // We assume that if we have a performance timer that the rAF callback
+    // gets a performance timer value. Not sure if this is always true.
+    const remaining = getTimeRemaining() - performance.now();
+    return remaining > 0 ? remaining : 0;
+  };
+} else {
+  timeRemaining = () => {
+    // Fallback to Date.now()
+    const remaining = getTimeRemaining() - Date.now();
+    return remaining > 0 ? remaining : 0;
+  };
+}
+
+const deadlineObject = {
+  timeRemaining,
+  didTimeout: false,
+};
+
+function ensureHostCallbackIsScheduled(highestPriorityNode) {
+  if (isPerformingWork) {
+    // Don't schedule work yet; wait until the next time we yield.
+    return;
+  }
+  const timesOutAt = highestPriorityNode.timesOutAt;
+  if (!isHostCallbackScheduled) {
+    isHostCallbackScheduled = true;
+  } else {
+    // Cancel the existing work.
+    cancelCallback();
+  }
+  // Schedule work using the highest priority callback's timeout.
+  requestCallback(flushWork, timesOutAt);
+}
+
+function computeAbsoluteTimeoutForPriority(currentTime, priority) {
+  if (priority === Deferred) {
+    return currentTime + DEFERRED_TIMEOUT;
+  }
+  throw new Error('Not yet implemented.');
+}
+
+function flushCallback(node) {
+  // This is already true; only assigning to appease Flow.
+  firstCallbackNode = node;
+
+  // Remove the node from the list before calling the callback. That way the
+  // list is in a consistent state even if the callback throws.
+  const next = firstCallbackNode.next;
+  if (firstCallbackNode === next) {
+    // This is the last callback in the list.
+    firstCallbackNode = null;
+  } else {
+    const previous = firstCallbackNode.previous;
+    firstCallbackNode = previous.next = next;
+    next.previous = previous;
+  }
+
+  node.next = node.previous = null;
+
+  // Now it's safe to call the callback.
+  const callback = node.callback;
+  callback(deadlineObject);
+}
+
+function flushWork(didTimeout) {
+  isPerformingWork = true;
+  deadlineObject.didTimeout = didTimeout;
+  try {
+    if (firstCallbackNode !== null) {
+      if (didTimeout) {
+        // Flush all the timed out callbacks without yielding.
+        do {
+          flushCallback(firstCallbackNode);
+        } while (
+          firstCallbackNode !== null &&
+          firstCallbackNode.timesOutAt <= getCurrentTime()
```

Comment:
Since `performance.now()` calls have a cost, I wonder if we could avoid recalculating it while we have expired work (since if 100 things expired checking the time is unnecessary — they can't "unexpire"). For example by having a nested loop. You read the time in the outer loop, and compare to it in the inner loop. Once you exit the inner loop, you read the time again. If there's nothing new that expired then we exit.